lib/fluent/compat/output.rb

```diff
@@ -337,22 +337,29 @@ def format_stream(tag, es)
       # because v0.12 BufferedOutput may overrides #format_stream, but original #handle_stream_simple method doesn't consider about it
       def handle_stream_simple(tag, es, enqueue: false)
         if @overrides_emit
-          current_emit_count = @emit_count
+          current_emit_count = @emit_count_metrics.get
           size = es.size
           key = data = nil
           begin
             emit(tag, es, NULL_OUTPUT_CHAIN)
             key, data = self.last_emit_via_buffer
           ensure
-            @emit_count = current_emit_count
+            @emit_count_metrics.set(current_emit_count)
             self.last_emit_via_buffer = nil
           end
           # on-the-fly key assignment can be done, and it's not configurable if Plugin#emit does it dynamically
           meta = @buffer.metadata(variables: (key && !key.empty? ? {key: key} : nil))
           write_guard do
             @buffer.write({meta => data}, format: ->(_data){ _data }, size: ->(){ size }, enqueue: enqueue)
           end
-          @counter_mutex.synchronize{ @emit_records += size }
+          if @enable_size_metrics
+            @counter_mutex.synchronize do
```

Comment:
I think `@counter_mutex` isn't needed any more because each `Metrics` object have a mutex lock.

Generally speaking, wrapping into same lock is desired as mentioned in https://github.com/fluent/fluentd/pull/3411/files/97d1692c05538de71b285fbaa374ff38ec443aa3#r647202429.
But keeping consistency of both value and using 2 locks aren't so meaningful in this case.
I think keeping the code simple is more important in this case.

```ruby
@emit_records_metrics.add(es.size)
@emit_size_metrics.add(es.to_msgpack_stream.bytesize) if @enable_size_metrics
```

Reply:
Fixed in https://github.com/fluent/fluentd/pull/3479/commits/aa79a4cd16bdab3190dd8fe8f5e2dc92a69c72ea.